File: src/main/java/gargoyle/ct/ui/impl/control/CTAction.java
package gargoyle.ct.ui.impl.control;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.SwingPropertyChangeSupport;
import java.beans.PropertyChangeListener;
import java.util.HashMap;
import java.util.Map;

@SuppressWarnings({"ConstructorNotProtectedInAbstractClass", "WeakerAccess"})
public abstract class CTAction implements Action {
    private static final String ENABLED = "enabled";
    private final Map<String, Object> values = new HashMap<>();
    private SwingPropertyChangeSupport changeSupport;

    public CTAction(String text) {
        setText(text);
        setToolTipText(text);
        setEnabled(true);
    }

    public CTAction(String text, Icon icon) {
        setText(text);
        setToolTipText(text);
        setIcon(icon);
        setEnabled(true);
    }

    public CTAction(String text, String tooltipText, Icon icon) {
        setText(text);
        setToolTipText(tooltipText);
        setIcon(icon);
        setEnabled(true);
    }

    public CTAction(String text, String tooltipText) {
        setText(text);
        setToolTipText(tooltipText);
        setEnabled(true);
    }

    protected CTAction() {
        setEnabled(true);
    }

    public void init(@NotNull AbstractButton menuItem) {
        menuItem.setAction(this);
        menuItem.setText(getText());
        menuItem.setToolTipText(getToolTipText());
        Icon icon = getIcon();
        if (icon != null) {
            menuItem.setIcon(icon);
        }
    }

    protected final String getText() {
        return String.valueOf(getValue(Action.NAME));
    }

    protected final void setText(String text) {
        putValue(Action.NAME, text);
    }

    protected final String getToolTipText() {
        return String.valueOf(getValue(Action.SHORT_DESCRIPTION));
    }

    protected final void setToolTipText(String text) {
        putValue(Action.SHORT_DESCRIPTION, text);
    }

    @Nullable
    protected final Icon getIcon() {
        return (Icon) getValue(Action.SMALL_ICON);
    }

    protected final void setIcon(Icon icon) {
        putValue(Action.SMALL_ICON, icon);
    }

    @Nullable
    @Override
    public final Object getValue(String key) {
        return values.get(key);
    }

    @Override
    public final void putValue(String key, Object value) {
        Object oldValue = getValue(key);
        values.put(key, value);
        firePropertyChange(key, oldValue, value);
    }

    @Override
    public final boolean isEnabled() {
        Object value = getValue(ENABLED);
        return value == null ? false : (Boolean) value;
    }

    @Override
    public final void setEnabled(boolean b) {
        putValue(ENABLED, b);
    }

    @SuppressWarnings("InstanceVariableUsedBeforeInitialized")
    protected final synchronized void firePropertyChange(String propertyName, @Nullable Object oldValue, @Nullable Object newValue) {
        if (changeSupport == null || oldValue != null && newValue != null && oldValue.equals(newValue)) {
            return;
        }
        changeSupport.firePropertyChange(propertyName, oldValue, newValue);
    }

    @Override
    public final synchronized void addPropertyChangeListener(PropertyChangeListener listener) {
        if (changeSupport == null) {
            changeSupport = new SwingPropertyChangeSupport(this);
        }
        changeSupport.addPropertyChangeListener(listener);
    }

    @Override
    public final synchronized void removePropertyChangeListener(PropertyChangeListener listener) {
        if (changeSupport == null) {
            return;
        }
        changeSupport.removePropertyChangeListener(listener);
    }
}

File: src/main/kotlin/gargoyle/ct/ui/impl/control/CTAction.kt
package gargoyle.ct.ui.impl.control

import java.beans.PropertyChangeListener
import javax.swing.AbstractButton
import javax.swing.Action
import javax.swing.Icon
import javax.swing.event.SwingPropertyChangeSupport

abstract class CTAction : Action {
    private val values: MutableMap<String, Any> = HashMap()
    private var changeSupport: SwingPropertyChangeSupport? = null

    constructor(text: String) {
        this.text = text
        toolTipText = text
        isEnabled = true
    }

    constructor(text: String, icon: Icon) {
        this.text = text
        toolTipText = text
        setIcon(icon)
        isEnabled = true
    }

    constructor(text: String, tooltipText: String, icon: Icon) {
        this.text = text
        toolTipText = tooltipText
        setIcon(icon)
        isEnabled = true
    }

    constructor(text: String, tooltipText: String) {
        this.text = text
        toolTipText = tooltipText
        isEnabled = true
    }

    protected constructor() {
        isEnabled = true
    }

    open fun init(menuItem: AbstractButton) {
        menuItem.action = this
        menuItem.text = text
        menuItem.toolTipText = toolTipText
        val icon = icon
        if (icon != null) {
            menuItem.icon = icon
        }
    }

    protected var text: String
        get() = getValue(Action.NAME).toString()
        protected set(text) {
            putValue(Action.NAME, text)
        }
    protected var toolTipText: String
        get() = getValue(Action.SHORT_DESCRIPTION).toString()
        protected set(text) {
            putValue(Action.SHORT_DESCRIPTION, text)
        }
    protected val icon: Icon?
        get() = getValue(Action.SMALL_ICON) as Icon?

    protected fun setIcon(icon: Icon) {
        putValue(Action.SMALL_ICON, icon)
    }

    override fun getValue(key: String): Any? {
        return values[key]
    }

    override fun putValue(key: String, value: Any) {
        val oldValue = getValue(key)
        values[key] = value
        firePropertyChange(key, oldValue, value)
    }

    override fun isEnabled(): Boolean {
        val value = getValue(ENABLED)
        return value != null && value as Boolean
    }

    override fun setEnabled(b: Boolean) {
        putValue(ENABLED, b)
    }

    @Synchronized
    protected fun firePropertyChange(propertyName: String?, oldValue: Any?, newValue: Any?) {
        if (changeSupport == null || oldValue != null && oldValue == newValue) {
            return
        }
        changeSupport!!.firePropertyChange(propertyName, oldValue, newValue)
    }

    @Synchronized
    override fun addPropertyChangeListener(listener: PropertyChangeListener) {
        if (changeSupport == null) {
            changeSupport = SwingPropertyChangeSupport(this)
        }
        changeSupport!!.addPropertyChangeListener(listener)
    }

    @Synchronized
    override fun removePropertyChangeListener(listener: PropertyChangeListener) {
        if (changeSupport == null) {
            return
        }
        changeSupport!!.removePropertyChangeListener(listener)
    }

    companion object {
        private const val ENABLED = "enabled"
    }
}
